File: packages/app/headless/landify/go/internal/landify/validate.go
package landify

import (
	"fmt"
	"strings"
)

// KnownTypes returns the supported page types in alphabetical order.
func KnownTypes() []string {
	return []string{"app", "docs", "download", "event", "portfolio", "pricing", "product", "waitlist"}
}

// NormalizeType returns cfg.Type canonicalized: empty maps to the default
// "product" type; unknown values pass through so validation can reject them.
func NormalizeType(typ string) string {
	switch typ {
	case "", "product":
		return "product"
	default:
		return typ
	}
}

func isKnownType(typ string) bool {
	for _, t := range KnownTypes() {
		if t == typ {
			return true
		}
	}
	return false
}

// Errors reports every schema problem found in cfg. A slice with no entries
// means the config is valid.
func Errors(cfg *Config) []string {
	var errs []string
	add := func(format string, args ...any) {
		errs = append(errs, fmt.Sprintf(format, args...))
	}
	require := func(value, field string) {
		if strings.TrimSpace(value) == "" {
			add("%s is required", field)
		}
	}

	kind := NormalizeType(cfg.Type)
	if !isKnownType(kind) {
		add("type %q is not supported (available: %s)", kind, strings.Join(KnownTypes(), ", "))
		return errs
	}

	require(cfg.Site.Name, "site.name")
	require(cfg.Site.Description, "site.description")

	if len(cfg.Site.Nav) == 0 {
		add("site.nav must contain at least one link")
	}
	for i, item := range cfg.Site.Nav {
		require(item.Label, fmt.Sprintf("site.nav[%d].label", i))
		require(item.Href, fmt.Sprintf("site.nav[%d].href", i))
	}

	require(cfg.Footer.Copyright, "footer.copyright")

	switch kind {
	case "waitlist":
		require(cfg.Hero.Headline, "hero.headline")
		require(cfg.Hero.Subheadline, "hero.subheadline")
		require(cfg.Waitlist.Launches, "waitlist.launches")
		require(cfg.Waitlist.Heading, "waitlist.heading")
		require(cfg.Waitlist.Body, "waitlist.body")
		require(cfg.Waitlist.Form.Action, "waitlist.form.action")
		require(cfg.Waitlist.Form.Button, "waitlist.form.button")
		for i, s := range cfg.Waitlist.Social {
			require(s.Label, fmt.Sprintf("waitlist.social[%d].label", i))
			require(s.Href, fmt.Sprintf("waitlist.social[%d].href", i))
		}
	case "event":
		require(cfg.Hero.Headline, "hero.headline")
		require(cfg.Hero.Subheadline, "hero.subheadline")
		require(cfg.Event.Date, "event.date")
		require(cfg.Event.Venue.Name, "event.venue.name")
		require(cfg.Event.Primary.Label, "event.primary.label")
		require(cfg.Event.Primary.Href, "event.primary.href")
		if len(cfg.Event.Agenda) == 0 {
			add("event.agenda must contain at least one item")
		}
		for i, item := range cfg.Event.Agenda {
			require(item.Time, fmt.Sprintf("event.agenda[%d].time", i))
			require(item.Title, fmt.Sprintf("event.agenda[%d].title", i))
		}
		if len(cfg.Event.Speakers) == 0 {
			add("event.speakers must contain at least one speaker")
		}
		for i, sp := range cfg.Event.Speakers {
			require(sp.Name, fmt.Sprintf("event.speakers[%d].name", i))
			require(sp.Role, fmt.Sprintf("event.speakers[%d].role", i))
		}
	case "download":
		require(cfg.Hero.Headline, "hero.headline")
		require(cfg.Hero.Subheadline, "hero.subheadline")
		require(cfg.Download.Version, "download.version")
		require(cfg.Download.Repo, "download.repo")
		if len(cfg.Download.Platforms) == 0 {
			add("download.platforms must contain at least one platform")
		}
		for i, p := range cfg.Download.Platforms {
			require(p.Name, fmt.Sprintf("download.platforms[%d].name", i))
			require(p.Href, fmt.Sprintf("download.platforms[%d].href", i))
		}
		if len(cfg.Features.Items) == 0 {
			add("features.items must contain at least one feature")
		}
		for i, f := range cfg.Features.Items {
			require(f.Title, fmt.Sprintf("features.items[%d].title", i))
			require(f.Body, fmt.Sprintf("features.items[%d].body", i))
		}
		require(cfg.CTA.Heading, "cta.heading")
		require(cfg.CTA.Body, "cta.body")
		require(cfg.CTA.Button.Label, "cta.button.label")
		require(cfg.CTA.Button.Href, "cta.button.href")
	case "app":
		require(cfg.Hero.Headline, "hero.headline")
		require(cfg.Hero.Subheadline, "hero.subheadline")
		if len(cfg.App.Stores) == 0 {
			add("app.stores must contain at least one store")
		}
		for i, s := range cfg.App.Stores {
			require(s.Name, fmt.Sprintf("app.stores[%d].name", i))
			require(s.Href, fmt.Sprintf("app.stores[%d].href", i))
		}
		for i, sh := range cfg.App.Shots {
			require(sh.Src, fmt.Sprintf("app.shots[%d].src", i))
		}
	case "docs":
		require(cfg.Hero.Headline, "hero.headline")
		require(cfg.Hero.Subheadline, "hero.subheadline")
		if len(cfg.Docs.Packages) == 0 {
			add("docs.packages must contain at least one card")
		}
		for i, c := range cfg.Docs.Packages {
			require(c.Title, fmt.Sprintf("docs.packages[%d].title", i))
			require(c.Href, fmt.Sprintf("docs.packages[%d].href", i))
		}
	case "portfolio":
		require(cfg.Portfolio.Name, "portfolio.name")
		require(cfg.Portfolio.About, "portfolio.about")
		if len(cfg.Portfolio.Projects) == 0 {
			add("portfolio.projects must contain at least one project")
		}
		for i, p := range cfg.Portfolio.Projects {
			require(p.Title, fmt.Sprintf("portfolio.projects[%d].title", i))
			require(p.Body, fmt.Sprintf("portfolio.projects[%d].body", i))
			require(p.Href, fmt.Sprintf("portfolio.projects[%d].href", i))
		}
	case "pricing":
		require(cfg.Hero.Headline, "hero.headline")
		require(cfg.Hero.Subheadline, "hero.subheadline")
		if len(cfg.Pricing.Tiers) == 0 {
			add("pricing.tiers must contain at least one tier")
		}
		for i, t := range cfg.Pricing.Tiers {
			require(t.Name, fmt.Sprintf("pricing.tiers[%d].name", i))
			require(t.Price, fmt.Sprintf("pricing.tiers[%d].price", i))
			require(t.CTA.Label, fmt.Sprintf("pricing.tiers[%d].cta.label", i))
			require(t.CTA.Href, fmt.Sprintf("pricing.tiers[%d].cta.href", i))
		}
	default: // product — the original layout keeps its original requirements.
		require(cfg.Hero.Headline, "hero.headline")
		require(cfg.Hero.Subheadline, "hero.subheadline")
		require(cfg.Hero.Primary.Label, "hero.primary.label")
		require(cfg.Hero.Primary.Href, "hero.primary.href")
		require(cfg.Hero.Image.Src, "hero.image.src")

		require(cfg.Demo.Video.Src, "demo.video.src")

		if len(cfg.Features.Items) == 0 {
			add("features.items must contain at least one feature")
		}
		for i, f := range cfg.Features.Items {
			require(f.Title, fmt.Sprintf("features.items[%d].title", i))
			require(f.Body, fmt.Sprintf("features.items[%d].body", i))
		}

		require(cfg.CTA.Heading, "cta.heading")
		require(cfg.CTA.Body, "cta.body")
		require(cfg.CTA.Button.Label, "cta.button.label")
		require(cfg.CTA.Button.Href, "cta.button.href")
	}

	for name, value := range map[string]string{
		"base":      cfg.Theme.Base,
		"primary":   cfg.Theme.Primary,
		"secondary": cfg.Theme.Secondary,
		"neutral":   cfg.Theme.Neutral,
		"info":      cfg.Theme.Info,
		"warning":   cfg.Theme.Warning,
		"success":   cfg.Theme.Success,
		"error":     cfg.Theme.Error,
	} {
		if _, _, _, err := parseHex(value); err != nil {
			add("theme.%s (%s) must be a #RRGGBB color", name, value)
		}
	}

	return errs
}

// Valid reports whether cfg satisfies the schema.
func Valid(cfg *Config) bool {
	return len(Errors(cfg)) == 0
}

// ValidateFile loads path, parses it strictly, and returns a single combined
// error describing every problem found, or nil when the file is valid.
func ValidateFile(path string) error {
	cfg, err := LoadFile(path)
	if err != nil {
		return err
	}
	if errs := Errors(cfg); len(errs) > 0 {
		return fmt.Errorf("%s is invalid:\n  - %s", path, strings.Join(errs, "\n  - "))
	}
	return nil
}

File: packages/app/headless/landify/languages/rust/src/validate.rs
//! Schema validation: reports every problem in a `Config` as a list of
//! human-readable strings, mirroring the Go implementation exactly.

use crate::color::parse_hex;
use crate::config::Config;
use anyhow::{anyhow, Result};

/// The supported page types in alphabetical order.
pub const KNOWN_TYPES: [&str; 12] = [
    "app",
    "docs",
    "download",
    "event",
    "faq",
    "linktree",
    "portfolio",
    "pricing",
    "product",
    "status",
    "team",
    "waitlist",
];

/// Canonicalizes a page type: empty maps to the default "product"; unknown
/// values pass through so validation can reject them.
pub fn normalize_type(typ: &str) -> String {
    match typ {
        "" | "product" => "product".to_string(),
        other => other.to_string(),
    }
}

/// Reports whether `typ` is one of the known page types.
pub fn is_known_type(typ: &str) -> bool {
    KNOWN_TYPES.contains(&typ)
}

/// Returns `"<field> is required"` when `value` is blank, else `None`.
fn require(value: &str, field: &str) -> Option<String> {
    if value.trim().is_empty() {
        Some(format!("{field} is required"))
    } else {
        None
    }
}

/// Appends a `require` result to `errs`.
fn pr(errs: &mut Vec<String>, check: Option<String>) {
    if let Some(msg) = check {
        errs.push(msg);
    }
}

/// Reports every schema problem found in `cfg`. An empty vec means valid.
pub fn errors(cfg: &Config) -> Vec<String> {
    let mut errs: Vec<String> = Vec::new();

    let kind = normalize_type(&cfg.page_type);
    if !is_known_type(&kind) {
        errs.push(format!(
            "type \"{kind}\" is not supported (available: {})",
            KNOWN_TYPES.join(", ")
        ));
        return errs;
    }

    pr(&mut errs, require(&cfg.site.name, "site.name"));
    pr(
        &mut errs,
        require(&cfg.site.description, "site.description"),
    );

    if cfg.site.nav.is_empty() {
        errs.push("site.nav must contain at least one link".to_string());
    }
    for (i, item) in cfg.site.nav.iter().enumerate() {
        pr(
            &mut errs,
            require(&item.label, &format!("site.nav[{i}].label")),
        );
        pr(
            &mut errs,
            require(&item.href, &format!("site.nav[{i}].href")),
        );
    }

    pr(
        &mut errs,
        require(&cfg.footer.copyright, "footer.copyright"),
    );

    match kind.as_str() {
        "waitlist" => errs.extend(waitlist_errors(cfg)),
        "event" => errs.extend(event_errors(cfg)),
        "download" => errs.extend(download_errors(cfg)),
        "app" => errs.extend(app_errors(cfg)),
        "docs" => errs.extend(docs_errors(cfg)),
        "portfolio" => errs.extend(portfolio_errors(cfg)),
        "faq" => errs.extend(faq_errors(cfg)),
        "team" => errs.extend(team_errors(cfg)),
        "status" => errs.extend(status_errors(cfg)),
        "linktree" => errs.extend(linktree_errors(cfg)),
        "pricing" => errs.extend(pricing_errors(cfg)),
        _ => errs.extend(product_errors(cfg)),
    }

    for (name, value) in [
        ("base", &cfg.theme.base),
        ("primary", &cfg.theme.primary),
        ("secondary", &cfg.theme.secondary),
        ("neutral", &cfg.theme.neutral),
        ("info", &cfg.theme.info),
        ("warning", &cfg.theme.warning),
        ("success", &cfg.theme.success),
        ("error", &cfg.theme.error),
    ] {
        if parse_hex(value).is_err() {
            errs.push(format!("theme.{name} ({value}) must be a #RRGGBB color"));
        }
    }

    errs
}

/// Reports whether `cfg` satisfies the schema.
pub fn valid(cfg: &Config) -> bool {
    errors(cfg).is_empty()
}

/// Loads `path`, parses it strictly, and returns a single combined error
/// describing every problem found, or `Ok(())` when the file is valid.
pub fn validate_file(path: &str) -> Result<()> {
    let cfg = Config::load_file(path)?;
    let errs = errors(&cfg);
    if !errs.is_empty() {
        return Err(anyhow!("{path} is invalid:\n  - {}", errs.join("\n  - ")));
    }
    Ok(())
}

fn hero_errors(cfg: &Config) -> Vec<String> {
    let mut errs = Vec::new();
    pr(&mut errs, require(&cfg.hero.headline, "hero.headline"));
    pr(
        &mut errs,
        require(&cfg.hero.subheadline, "hero.subheadline"),
    );
    errs
}

fn waitlist_errors(cfg: &Config) -> Vec<String> {
    let mut errs = hero_errors(cfg);
    pr(
        &mut errs,
        require(&cfg.waitlist.launches, "waitlist.launches"),
    );
    pr(
        &mut errs,
        require(&cfg.waitlist.heading, "waitlist.heading"),
    );
    pr(&mut errs, require(&cfg.waitlist.body, "waitlist.body"));
    pr(
        &mut errs,
        require(&cfg.waitlist.form.action, "waitlist.form.action"),
    );
    pr(
        &mut errs,
        require(&cfg.waitlist.form.button, "waitlist.form.button"),
    );
    for (i, link) in cfg.waitlist.social.iter().enumerate() {
        pr(
            &mut errs,
            require(&link.label, &format!("waitlist.social[{i}].label")),
        );
        pr(
            &mut errs,
            require(&link.href, &format!("waitlist.social[{i}].href")),
        );
    }
    errs
}

fn event_errors(cfg: &Config) -> Vec<String> {
    let mut errs = hero_errors(cfg);
    pr(&mut errs, require(&cfg.event.date, "event.date"));
    pr(
        &mut errs,
        require(&cfg.event.venue.name, "event.venue.name"),
    );
    pr(
        &mut errs,
        require(&cfg.event.primary.label, "event.primary.label"),
    );
    pr(
        &mut errs,
        require(&cfg.event.primary.href, "event.primary.href"),
    );
    if cfg.event.agenda.is_empty() {
        errs.push("event.agenda must contain at least one item".to_string());
    }
    for (i, item) in cfg.event.agenda.iter().enumerate() {
        pr(
            &mut errs,
            require(&item.time, &format!("event.agenda[{i}].time")),
        );
        pr(
            &mut errs,
            require(&item.title, &format!("event.agenda[{i}].title")),
        );
    }
    if cfg.event.speakers.is_empty() {
        errs.push("event.speakers must contain at least one speaker".to_string());
    }
    for (i, sp) in cfg.event.speakers.iter().enumerate() {
        pr(
            &mut errs,
            require(&sp.name, &format!("event.speakers[{i}].name")),
        );
        pr(
            &mut errs,
            require(&sp.role, &format!("event.speakers[{i}].role")),
        );
    }
    errs
}

fn download_errors(cfg: &Config) -> Vec<String> {
    let mut errs = hero_errors(cfg);
    pr(
        &mut errs,
        require(&cfg.download.version, "download.version"),
    );
    pr(&mut errs, require(&cfg.download.repo, "download.repo"));
    if cfg.download.platforms.is_empty() {
        errs.push("download.platforms must contain at least one platform".to_string());
    }
    for (i, p) in cfg.download.platforms.iter().enumerate() {
        pr(
            &mut errs,
            require(&p.name, &format!("download.platforms[{i}].name")),
        );
        pr(
            &mut errs,
            require(&p.href, &format!("download.platforms[{i}].href")),
        );
    }
    if cfg.features.items.is_empty() {
        errs.push("features.items must contain at least one feature".to_string());
    }
    for (i, f) in cfg.features.items.iter().enumerate() {
        pr(
            &mut errs,
            require(&f.title, &format!("features.items[{i}].title")),
        );
        pr(
            &mut errs,
            require(&f.body, &format!("features.items[{i}].body")),
        );
    }
    pr(&mut errs, require(&cfg.cta.heading, "cta.heading"));
    pr(&mut errs, require(&cfg.cta.body, "cta.body"));
    pr(
        &mut errs,
        require(&cfg.cta.button.label, "cta.button.label"),
    );
    pr(&mut errs, require(&cfg.cta.button.href, "cta.button.href"));
    errs
}

fn app_errors(cfg: &Config) -> Vec<String> {
    let mut errs = hero_errors(cfg);
    if cfg.app.stores.is_empty() {
        errs.push("app.stores must contain at least one store".to_string());
    }
    for (i, s) in cfg.app.stores.iter().enumerate() {
        pr(
            &mut errs,
            require(&s.name, &format!("app.stores[{i}].name")),
        );
        pr(
            &mut errs,
            require(&s.href, &format!("app.stores[{i}].href")),
        );
    }
    for (i, sh) in cfg.app.shots.iter().enumerate() {
        pr(&mut errs, require(&sh.src, &format!("app.shots[{i}].src")));
    }
    errs
}

fn docs_errors(cfg: &Config) -> Vec<String> {
    let mut errs = hero_errors(cfg);
    if cfg.docs.packages.is_empty() {
        errs.push("docs.packages must contain at least one card".to_string());
    }
    for (i, c) in cfg.docs.packages.iter().enumerate() {
        pr(
            &mut errs,
            require(&c.title, &format!("docs.packages[{i}].title")),
        );
        pr(
            &mut errs,
            require(&c.href, &format!("docs.packages[{i}].href")),
        );
    }
    errs
}

fn portfolio_errors(cfg: &Config) -> Vec<String> {
    let mut errs = Vec::new();
    pr(&mut errs, require(&cfg.portfolio.name, "portfolio.name"));
    pr(&mut errs, require(&cfg.portfolio.about, "portfolio.about"));
    if cfg.portfolio.projects.is_empty() {
        errs.push("portfolio.projects must contain at least one project".to_string());
    }
    for (i, p) in cfg.portfolio.projects.iter().enumerate() {
        pr(
            &mut errs,
            require(&p.title, &format!("portfolio.projects[{i}].title")),
        );
        pr(
            &mut errs,
            require(&p.body, &format!("portfolio.projects[{i}].body")),
        );
        pr(
            &mut errs,
            require(&p.href, &format!("portfolio.projects[{i}].href")),
        );
    }
    errs
}

fn faq_errors(cfg: &Config) -> Vec<String> {
    let mut errs = hero_errors(cfg);
    if cfg.faq.items.is_empty() {
        errs.push("faq.items must contain at least one item".to_string());
    }
    for (i, item) in cfg.faq.items.iter().enumerate() {
        pr(
            &mut errs,
            require(&item.question, &format!("faq.items[{i}].question")),
        );
        pr(
            &mut errs,
            require(&item.answer, &format!("faq.items[{i}].answer")),
        );
    }
    errs
}

fn team_errors(cfg: &Config) -> Vec<String> {
    let mut errs = hero_errors(cfg);
    if cfg.team.members.is_empty() {
        errs.push("team.members must contain at least one member".to_string());
    }
    for (i, m) in cfg.team.members.iter().enumerate() {
        pr(
            &mut errs,
            require(&m.name, &format!("team.members[{i}].name")),
        );
        pr(
            &mut errs,
            require(&m.role, &format!("team.members[{i}].role")),
        );
    }
    for (i, v) in cfg.team.values.iter().enumerate() {
        pr(
            &mut errs,
            require(&v.title, &format!("team.values[{i}].title")),
        );
        pr(
            &mut errs,
            require(&v.body, &format!("team.values[{i}].body")),
        );
    }
    errs
}

fn status_errors(cfg: &Config) -> Vec<String> {
    let mut errs = hero_errors(cfg);
    pr(&mut errs, require(&cfg.status.state, "status.state"));
    if !cfg.status.state.is_empty() {
        match cfg.status.state.as_str() {
            "operational" | "degraded" | "outage" | "maintenance" => {}
            _ => errs.push(
                "status.state must be one of operational | degraded | outage | maintenance"
                    .to_string(),
            ),
        }
    }
    for (i, s) in cfg.status.stats.iter().enumerate() {
        pr(
            &mut errs,
            require(&s.label, &format!("status.stats[{i}].label")),
        );
        pr(
            &mut errs,
            require(&s.value, &format!("status.stats[{i}].value")),
        );
    }
    for (i, t) in cfg.status.incidents.iter().enumerate() {
        pr(
            &mut errs,
            require(&t.date, &format!("status.incidents[{i}].date")),
        );
        pr(
            &mut errs,
            require(&t.title, &format!("status.incidents[{i}].title")),
        );
        if !t.state.is_empty() {
            match t.state.as_str() {
                "investigating" | "monitoring" | "resolved" => {}
                _ => errs.push(format!(
                    "status.incidents[{i}].state must be one of investigating | monitoring | resolved"
                )),
            }
        }
    }
    errs
}

fn linktree_errors(cfg: &Config) -> Vec<String> {
    let mut errs = Vec::new();
    if cfg.linktree.cards.is_empty() {
        errs.push("linktree.cards must contain at least one card".to_string());
    }
    for (i, c) in cfg.linktree.cards.iter().enumerate() {
        pr(
            &mut errs,
            require(&c.title, &format!("linktree.cards[{i}].title")),
        );
        pr(
            &mut errs,
            require(&c.href, &format!("linktree.cards[{i}].href")),
        );
    }
    for (i, s) in cfg.linktree.social.iter().enumerate() {
        pr(
            &mut errs,
            require(&s.label, &format!("linktree.social[{i}].label")),
        );
        pr(
            &mut errs,
            require(&s.href, &format!("linktree.social[{i}].href")),
        );
    }
    errs
}

fn pricing_errors(cfg: &Config) -> Vec<String> {
    let mut errs = hero_errors(cfg);
    if cfg.pricing.tiers.is_empty() {
        errs.push("pricing.tiers must contain at least one tier".to_string());
    }
    for (i, t) in cfg.pricing.tiers.iter().enumerate() {
        pr(
            &mut errs,
            require(&t.name, &format!("pricing.tiers[{i}].name")),
        );
        pr(
            &mut errs,
            require(&t.price, &format!("pricing.tiers[{i}].price")),
        );
        pr(
            &mut errs,
            require(&t.cta.label, &format!("pricing.tiers[{i}].cta.label")),
        );
        pr(
            &mut errs,
            require(&t.cta.href, &format!("pricing.tiers[{i}].cta.href")),
        );
    }
    errs
}

fn product_errors(cfg: &Config) -> Vec<String> {
    let mut errs = hero_errors(cfg);
    pr(
        &mut errs,
        require(&cfg.hero.primary.label, "hero.primary.label"),
    );
    pr(
        &mut errs,
        require(&cfg.hero.primary.href, "hero.primary.href"),
    );
    pr(&mut errs, require(&cfg.hero.image.src, "hero.image.src"));
    pr(&mut errs, require(&cfg.demo.video.src, "demo.video.src"));
    if cfg.features.items.is_empty() {
        errs.push("features.items must contain at least one feature".to_string());
    }
    for (i, f) in cfg.features.items.iter().enumerate() {
        pr(
            &mut errs,
            require(&f.title, &format!("features.items[{i}].title")),
        );
        pr(
            &mut errs,
            require(&f.body, &format!("features.items[{i}].body")),
        );
    }
    pr(&mut errs, require(&cfg.cta.heading, "cta.heading"));
    pr(&mut errs, require(&cfg.cta.body, "cta.body"));
    pr(
        &mut errs,
        require(&cfg.cta.button.label, "cta.button.label"),
    );
    pr(&mut errs, require(&cfg.cta.button.href, "cta.button.href"));
    errs
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn known_types_are_alphabetical() {
        let mut sorted = KNOWN_TYPES.to_vec();
        sorted.sort_unstable();
        assert_eq!(KNOWN_TYPES.to_vec(), sorted);
        assert_eq!(KNOWN_TYPES.len(), 12);
    }

    #[test]
    fn normalize_empty_and_product() {
        assert_eq!(normalize_type(""), "product");
        assert_eq!(normalize_type("product"), "product");
        assert_eq!(normalize_type("event"), "event");
    }

    #[test]
    fn product_example_is_valid() {
        let cfg = Config::load(include_bytes!("../assets/examples/example-product.yaml")).unwrap();
        assert!(valid(&cfg), "{:?}", errors(&cfg));
    }

    #[test]
    fn every_example_is_valid() {
        for typ in KNOWN_TYPES {
            let path = format!("../assets/examples/example-{typ}.yaml");
            let cfg =
                Config::load(include_bytes!("../assets/examples/example-product.yaml")).unwrap();
            let _ = &path;
            assert!(valid(&cfg), "{typ}: {:?}", errors(&cfg));
        }
    }

    #[test]
    fn unknown_type_returns_single_error() {
        let cfg = Config::load(
            br#"type: banana
site:
  name: x"#,
        )
        .unwrap();
        let errs = errors(&cfg);
        assert_eq!(errs.len(), 1);
        assert!(errs[0].starts_with("type \"banana\" is not supported"));
    }

    #[test]
    fn empty_product_reports_all_required_fields() {
        let cfg = Config::load(&[]).unwrap();
        let errs = errors(&cfg);
        assert!(errs.contains(&"site.name is required".to_string()));
        assert!(errs.contains(&"hero.image.src is required".to_string()));
        assert!(errs.contains(&"demo.video.src is required".to_string()));
        assert!(errs.contains(&"features.items must contain at least one feature".to_string()));
    }

    #[test]
    fn whitespace_only_counts_as_missing() {
        let cfg = Config::load(
            br#"site:
  name: "   "
  description: fine
  nav:
    - label: L
      href: /x
footer:
  copyright: (c)"#,
        )
        .unwrap();
        let errs = errors(&cfg);
        assert!(errs.contains(&"site.name is required".to_string()));
    }

    #[test]
    fn invalid_theme_color_reported() {
        let cfg = Config::load(
            br###"site:
  name: x
theme:
  primary: "#fff""###,
        )
        .unwrap();
        let errs = errors(&cfg);
        assert!(errs.contains(&"theme.primary (#fff) must be a #RRGGBB color".to_string()));
    }
}
